extensions/CHCAT1320/midi.js

```diff
@@ -0,0 +1,111 @@
+//# sourceMappingURL=Midi.js.map
+`;
+
+(function (Scratch) {
+  class MidiParserExtension {
+    constructor(runtime) {
+      this.runtime = runtime;
+      this.notesData = [];
+      this.errorMessage = '';
+      this.loadMidiLibrary();
+    }
+
+    loadMidiLibrary() {
+      try {
+        const script = document.createElement('script');
+        script.textContent = MidiLibrary; // Embedded library code
+        document.head.appendChild(script);
+      } catch (error) {
+        this.errorMessage = 'Failed to load MIDI library: ' + error.message;
+      }
+    }
+
+    getInfo() {
+      return {
+        id: 'midiParser',
+        name: 'MIDI Parser',
+        color1: '#4C97FF',
+        color2: '#337BCC',
+        color3: '#2C6CA3',
+        blocks: [
+          {
+            opcode: 'parseMidiDataUrl',
+            blockType: Scratch.BlockType.REPORTER,
+            text: 'Parse MIDI Data URL [DATA_URL]',
+            arguments: {
+              DATA_URL: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: ''
+              }
+            }
+          }
+        ]
+      };
+    }
+
+    parseMidiDataUrl(args) {
+      this.errorMessage = '';
+      if (!window.Midi) {
+        return 'MIDI library failed to load';
+      }
+
+      const dataUrl = args.DATA_URL.trim();
+      if (!dataUrl) {
+        return 'MIDI Data URL is empty';
+      }
+
+      try {
+        // Parse Data URL
+        const base64Data = dataUrl.split(',')[1];
+        if (!base64Data) {
+          throw new Error('Invalid Data URL format');
+        }
+
+        // Decode Base64 data to ArrayBuffer
+        const binaryString = atob(base64Data);
+        const len = binaryString.length;
+        const buffer = new ArrayBuffer(len);
+        const view = new Uint8Array(buffer);
+        for (let i = 0; i < len; i++) {
+          view[i] = binaryString.charCodeAt(i);
+        }
+
+        // Parse MIDI file
+        const midi = new window.Midi(buffer);
+
+        // Store note information
+        this.notesData = [];
+        midi.tracks.forEach((track, trackIndex) => {
+          track.notes.forEach(note => {
+            this.notesData.push({
+              track: trackIndex,
```

Comment:
Suggestion: consider outputting this as `trackIndex + 1` since scratch indexes start at `1` instead of `0`